### ISSUE TITLE
Possible bug in  captured_path.py

### ISSUE THREAD

**Issue by egoltsman** [ggonnella/gfapy#18]
Hi,
I ran into a weird issue where when processing an oriented path from a GFA2 file,  captured_path.py throws an exception claiming that the path is invalid:
```
Traceback (most recent call last):
  File "add_annotation_to_gfa.py", line 159, in <module>
    nFeatAdded = add_features(gfa2, p.captured_segments)
  File "/global/homes/e/eugeneg/.conda/envs/panG/lib/python3.8/site-packages/gfapy/line/common/dynamic_fields.py", line 19, in __getattribute__
    attr = super().__getattribute__(name)
  File "/global/homes/e/eugeneg/.conda/envs/panG/lib/python3.8/site-packages/gfapy/line/group/ordered/captured_path.py", line 7, in captured_segments
    return [ x for x in self.captured_path if isinstance(x.line, gfapy.line.segment.GFA2) ]
  File "/global/homes/e/eugeneg/.conda/envs/panG/lib/python3.8/site-packages/gfapy/line/common/dynamic_fields.py", line 19, in __getattribute__
    attr = super().__getattribute__(name)
  File "/global/homes/e/eugeneg/.conda/envs/panG/lib/python3.8/site-packages/gfapy/line/group/ordered/captured_path.py", line 20, in captured_path
    return self._compute_captured_path()[0]
  File "/global/homes/e/eugeneg/.conda/envs/panG/lib/python3.8/site-packages/gfapy/line/group/ordered/captured_path.py", line 26, in _compute_captured_path
    path, prev_edge = self._push_item_on_se_path(path, prev_edge, item)
  File "/global/homes/e/eugeneg/.conda/envs/panG/lib/python3.8/site-packages/gfapy/line/group/ordered/captured_path.py", line 52, in _push_item_on_se_path
    self._push_nonfirst_edge_on_se_path(path, item)
  File "/global/homes/e/eugeneg/.conda/envs/panG/lib/python3.8/site-packages/gfapy/line/group/ordered/captured_path.py", line 138, in _push_nonfirst_edge_on_se_path
    raise gfapy.NotFoundError(
gfapy.error.NotFoundError: Path is not valid, elements are not contiguous
```

I checked the path, and the elements are in fact contiguous and a unique edge exists for every pair of path segments.  Looking at the code it seems that the culprit is the call to invert the orientation of the spanning segments when adding a "negative" edge:

https://github.com/ggonnella/gfapy/blob/29063e79a85992dec385c661c6224997ba65b954/gfapy/line/group/ordered/captured_path.py#L132

Somehow, in my case, an edge was being visited twice (not sure how that's possible since all of the path elements are unique),  and since the invert() method flips the orientation *in place*, it would effectively undo itself, i.e.  flipping - to + on the first encounter and then - back to + the second time around.   This results in the mismatch between prev_os and possible_prev[i] , causing the error.  
I was able to get around this by replacing 
possible_prev[i].invert()
with 
possible_prev[i] = possible_prev[i].inverted() 

(inverted() apparently returns a local copy)

I didn't want to make a pull request because I'm not sure I've addressed the root cause, which I think is the fact that there is a second visit to the same edge on a single path.   I can see cases where this would be true, e.g. when a path involves a cycle, but I have no such paths in my graph afaik.   Perhaps there is a thread locking issue during parallel path walking? Just a wild guess...

Thanks

**Comment by egoltsman** [comment]
p.s.  This doesn't occur when processing the same graph in GFA1 format as there are no oriented edges in the path to worry about.

**Comment by egoltsman** [comment]
Sorry, scratch the last comment. It doesn't look like there's multi-threading anywhere in the code.
Perhaps the way I'm using the data structures in my own code has something to do with this.  Here's the snippet of my code that builds the gfapy object and grabs the  .captured_segments  list for every path.  In the process I add Fragment lines to some of the segments on the path but don't modify the path at all.  At the end the updated graph is printed out.
```
gfa=gfapy.Gfa.from_file(inGfaFile)                                                                                                                                                                                                                            
                                                                                                                                                                                                                                                              
if gfa.version == 'gfa1':                                                                                                                                                                                                                                     
    print('\tGFA1 format detected from input')                                                                                                                                                                                                                
    gfa2 = gfa.to_gfa2()                                                                                                                                                                                                                                      
elif gfa.version == 'gfa2':                                                                                                                                                                                                                                   
    print('\tGFA2 format detected from input')                                                                                                                                                                                                                
    gfa2 = gfa                                                                                                                                                                                                                                                
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               
for p in gfa2.paths:                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     
    if (p is blah):                                                                                                                                                                                                                                                                                                                                                                                                                                                              
        nFeatAdded = add_features(gfa2, p.captured_segments)                                                                                                                                                                                                  
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           
gfa2.to_file(outGfaFile)   
```

**Comment by ggonnella** [comment]
Hi @egoltsman,

thank you very much for the bug report. Despite having a bit forgotten how I wrote that part, I had a look to the code again, and I think you are right.

The `invert()` call is modifying the oriented line, and instead `inverted()` should be used at that place. I commited that.

It would be nice to have a test covering this, which I do not have yet. Is it perhaps possible for you to provide a failing case for the previous version?

Best regards,
Giorgio
